### PR TITLE
fix(web): users should be onboarded once

### LINKED ISSUE #141: onboarding flow is shown on each sign in
- sign in for the first time
- onboarding flow is shown (expected)
- clear cookies
- sign in with the same account again
- onboarding flow is shown again

my guess is hasOnboarded needs to be a field in the `user` table

### PR DESCRIPTION
**Pull request by aryasaatvik**
# Fix(web): Users should be onboarded once

## Overview
This pull request fixes an issue where users were being redirected to the onboarding page multiple times, even after completing the onboarding process. The changes ensure that users are only onboarded once and are then redirected to the main application.

Resolves #141 

## Changes

- Key Changes:
  - Added a new `hasOnboarded` field to the `users` table in the database schema.
  - Implemented a `completeOnboarding` function in the `doers.ts` file to update the `hasOnboarded` field for the current user.
  - Updated the `layout.ts` file to check if the user has completed onboarding before rendering the main application.
  - Removed the `firstTime` query parameter from the `home` page and the associated logic.

- New Features:
  - None

- Refactoring:
  - Moved the user authentication and onboarding check logic from the `home` page to the `layout.ts` file, improving separation of concerns.
  - Consolidated the user onboarding logic into a single `completeOnboarding` function, making it more reusable and maintainable.

The main purpose of this pull request is to ensure that users are only onboarded once, preventing unnecessary redirects to the onboarding page. This improves the user experience and ensures that users can access the main application functionality after completing the onboarding process.

> ✨ Generated with love by Kaizen ❤️

<details>
<summary>Original Description</summary>
None
</details>

### CODE REVIEW

**Comment by aryasaatvik** [comment]
> ## Code Review
> ❗ **Attention Required:** This PR has potential issues. 🚨
> 
> ### Database Migration
> --> The new column 'hasOnboarded' is defined as an integer but should be a boolean type to represent true/false values.
> ### Schema Definition
> --> The 'hasOnboarded' field is defined as an integer with a default value of false, which is incorrect.
> > ✨ Generated with love by Kaizen ❤️
> 
> Useful Commands

sqlite doesn't have a boolean data type

**Comment by aryasaatvik** [comment]
> ## Code Review
> ❗ **Attention Required:** This PR has potential issues. 🚨
> 
> ### Database Migration
> --> The new column 'hasOnboarded' is defined as an integer but should be a boolean type to represent true/false values.
> ### Schema Definition
> --> The 'hasOnboarded' field is defined as an integer with a default value of false, which is incorrect.
> > ✨ Generated with love by Kaizen ❤️
> 
> Useful Commands
> * **Feedback:** Reply with `!feedback [your message]`
> * **Review:** Reply with `!review`

!feedback sqlite doesn't have a boolean data type

**Comment by aryasaatvik** [comment]
@Dhravya running this query after the migration should set the hasOnboarded boolean retroactively for users

```sql
-- count number of rows in storedContent for each userId
-- if count is > 0, then set hasOnboarded to true for that userId in user table

WITH contentCount AS (
  SELECT userId, COUNT(*) AS count
  FROM storedContent
  GROUP BY userId
)
UPDATE user
SET hasOnboarded = 1
WHERE id IN (SELECT userId FROM contentCount WHERE count > 0);
```

File: apps/web/migrations/0001_mighty_newton_destine.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+ALTER TABLE `user` ADD `hasOnboarded` integer DEFAULT false;
```
[the remark this answers is not in the document]
Comment (aryasaatvik) [reply apps/web/migrations/0001_mighty_newton_destine.sql]: nope https://github.com/supermemoryai/supermemory/blob/main/apps/web/migrations/0000_setup.sql/#L103

File: apps/web/server/db/schema.ts [hunk 2]
```diff
@@ -22,6 +22,7 @@ export const users = createTable(
 		emailVerified: integer("emailVerified", { mode: "timestamp_ms" }),
 		image: text("image"),
 		telegramId: text("telegramId"),
+		hasOnboarded: integer("hasOnboarded", { mode: "boolean" }).default(false),
```
Comment (aryasaatvik) [inline apps/web/server/db/schema.ts:25]: sqlite doesn't have boolen type

File: apps/web/migrations/0001_mighty_newton_destine.sql (hunk 1 above)
Comment (aryasaatvik) [inline apps/web/migrations/0001_mighty_newton_destine.sql]: @Dhravya should this be a part of setup? should the meta files also be commited?

Comment (Dhravya) [reply apps/web/migrations/0001_mighty_newton_destine.sql]: Yes, this and the meta file both are a part of the commit 

this allows other contributors to run wrangler d1 migrations apply command

**Comment by Dhravya** [comment]
This command worked for updating the db
```
UPDATE user
SET hasOnboarded = 1
WHERE id IN (
    SELECT user
    FROM storedContent
    GROUP BY user
    HAVING COUNT(id) > 1
);
```

**Comment by Dhravya** [comment]
✅ Made the prod change

**Review by Dhravya — APPROVED** [review]
fabulous work :)

**Comment by Dhravya** [comment]
Merged. Thanks @aryasaatvik !

**Pull request merged**